server/src/main/java/org/spine3/server/entity/VisibilityGuard.java

```diff
@@ -0,0 +1,182 @@
+import org.spine3.option.EntityOptions;
+import org.spine3.type.TypeName;
+
+import javax.annotation.Nullable;
+import java.util.Collection;
+import java.util.Map;
+import java.util.Set;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+import static com.google.common.collect.Iterables.transform;
+import static com.google.common.collect.Maps.filterValues;
+import static com.google.common.collect.Maps.newHashMap;
+import static org.spine3.util.Exceptions.newIllegalArgumentException;
+import static org.spine3.util.Exceptions.newIllegalStateException;
+
+/**
+ * A registry of repositories that controls access to them depending on the visibility of
+ * corresponding entity states.
+ *
+ * @author Alexander Yevsyukov
+ * @see EntityOptions
+ */
+@Internal
+public final class VisibilityGuard {
+
+    private final Map<Class, RepositoryAccess> repositories = newHashMap();
+
+    private VisibilityGuard() {
+        // Prevent instantiation from outside.
+    }
+
+    /**
+     * Creates a new instance of the guard.
+     */
+    public static VisibilityGuard newInstance() {
+        return new VisibilityGuard();
+    }
+
+    /**
+     * Registers the passed repository with the guard.
+     */
+    public void register(Repository repository) {
+        checkNotNull(repository);
+        final Class stateClass = repository.getEntityStateClass();
+        checkNotAlreadyRegistered(stateClass);
+        repositories.put(stateClass, new RepositoryAccess(repository));
+    }
+
+    private void checkNotAlreadyRegistered(Class stateClass) {
+        final RepositoryAccess alreadyRegistered = repositories.get(stateClass);
+        if (alreadyRegistered != null) {
+            throw newIllegalStateException(
+                    "A repository for the state class %s already registered: %s",
+                    stateClass, alreadyRegistered);
+        }
+    }
+
+    /**
+     * Verifies if there is a registered repository for the passed entity state class.
+     */
+    public boolean hasRepository(Class<? extends Message> stateClass) {
+        final boolean result = repositories.containsKey(stateClass);
+        return result;
+    }
+
+    /**
+     * Obtains the repository for the passed entity state class.
+     *
+     * @param stateClass the class of the state of entities managed by the repository
+     * @return the repository wrapped into {@code Optional} or {@code Optional#absent()} if the
+     * entity state is {@linkplain Visibility#NONE not visible}
+     * @throws IllegalArgumentException if the repository for the passed state class was not
+     *                                  {@linkplain #register(Repository) registered} with the guard
+     *                                  prior to this call, or if all repositories were
+     *                                  {@linkplain #shutDownRepositories() shut down}
+     */
+    public Optional<Repository> getRepository(Class<? extends Message> stateClass) {
```

Comment:
As long we we perform the `checkNotNull(...)` in `getEntityTypes(visibility)` below, I think we should do the same for `stateClass` as well.